core-java/src/main/java/org/spine3/server/procman/ProcessManager.java

```diff
@@ -0,0 +1,288 @@
+
+/**
+ * A central processing unit used to maintain the state of the sequence and determine the next processing step
+ * based on intermediate results.
+ * <p>
+ * An independent component that reacts to domain events in a cross-aggregate, eventually consistent manner.
+ * Used for externalizing the decisions on the logic flow from the business logic.
+ * <p>
+ * Event/command handlers are invoked by the {@link ProcessManagerRepository}
+ * that manages instances of a process manager class.
+ * <p>
+ * There is a common confusion between Process Managers and Sagas.
+ * See <a href="http://kellabyte.com/2012/05/30/clarifying-the-saga-pattern/">this</a> and
+ * <a href="https://dzone.com/articles/are-sagas-and-workflows-same-t">this topic</a>
+ * to understand the difference between them.
+ *
+ * @param <I> the type of the process manager IDs
+ * @param <M> the type of the process manager state
+ * @see <a href="http://www.enterpriseintegrationpatterns.com/patterns/messaging/ProcessManager.html">Process Manager Pattern</a>
+ * @see <a href="https://msdn.microsoft.com/en-us/library/jj591569.aspx">CQRS Journey Guide: A Saga on Sagas</a>
+ * @author Alexander Litus
+ */
+public abstract class ProcessManager<I, M extends Message> extends Entity<I, M> {
+
+    /**
+     * Keeps initialization state of the process manager.
+     */
+    private volatile boolean initialized = false;
+
+    /**
+     * The map of command handler methods for this process manager.
+     *
+     * @see Registry
+     */
+    private MethodMap commandHandlers;
+
+    /**
+     * The map of event handlers for this process manager.
+     *
+     * @see Registry
+     */
+    private MethodMap eventHandlers;
+
+    protected ProcessManager(I id) {
+        super(id);
+    }
+
+    /**
+     * Performs initialization of the instance and registers this class of process managers
+     * in the {@link Registry} if it is not registered yet.
+     */
+    private void init() {
+        if (!this.initialized) {
+            final Registry registry = Registry.getInstance();
+            final Class<? extends ProcessManager> pmClass = getClass();
+            if (!registry.contains(pmClass)) {
+                registry.register(pmClass);
+            }
+            commandHandlers = registry.getCommandHandlers(pmClass);
+            eventHandlers = registry.getEventHandlers(pmClass);
+            this.initialized = true;
+        }
+    }
+
+    /**
+     * Dispatches a command to the command handler method of the process manager.
+     *
+     * @param command the command to be executed on the process manager
+     * @param context of the command
+     * @throws InvocationTargetException if an exception occurs during command dispatching
+     */
+    protected List<EventRecord> dispatchCommand(Message command, CommandContext context) throws InvocationTargetException {
+        checkNotNull(command, "command is null");
+        checkNotNull(context, "command context is null");
```

Comment:
Just "context". We have information about nullity in the fact of having NPE in the stack trace.